File: mcp-sentinel-python/src/mcp_sentinel/detectors/secrets.py
"""
Secrets detector for finding hardcoded credentials and API keys.
"""

import re
from typing import List, Dict, Pattern
from pathlib import Path

from mcp_sentinel.detectors.base import BaseDetector
from mcp_sentinel.models.vulnerability import (
    Vulnerability,
    VulnerabilityType,
    Severity,
    Confidence,
)


class SecretsDetector(BaseDetector):
    """
    Detector for hardcoded secrets, API keys, and credentials.

    Detects 15+ types of secrets:
    - AWS keys
    - OpenAI/Anthropic API keys
    - Private keys (RSA, EC, SSH)
    - JWT tokens
    - Database credentials
    - GitHub tokens
    - And more...
    """

    def __init__(self):
        """Initialize the secrets detector."""
        super().__init__(name="SecretsDetector", enabled=True)
        self.patterns: Dict[str, Pattern] = self._compile_patterns()

    def _compile_patterns(self) -> Dict[str, Pattern]:
        """Compile regex patterns for secret detection."""
        return {
            # AWS Access Keys
            "aws_access_key": re.compile(r"(A3T[A-Z0-9]|AKIA|AGPA|AIDA|AROA|AIPA|ANPA|ANVA|ASIA)[A-Z0-9]{16}"),

            # AWS Secret Keys
            "aws_secret_key": re.compile(r"(?i)aws(.{0,20})?['\"][0-9a-zA-Z/+]{40}['\"]"),

            # OpenAI API Keys
            "openai_api_key": re.compile(r"sk-[a-zA-Z0-9]{48}"),
            "openai_legacy_key": re.compile(r"sk-[a-zA-Z0-9]{32}"),

            # Anthropic Claude API Keys
            "anthropic_api_key": re.compile(r"sk-ant-api03-[a-zA-Z0-9\-_]{95,}"),

            # Google API Keys
            "google_api_key": re.compile(r"AIza[0-9A-Za-z\-_]{35}"),

            # GitHub Personal Access Tokens
            "github_token": re.compile(r"ghp_[0-9a-zA-Z]{36}"),
            "github_oauth": re.compile(r"gho_[0-9a-zA-Z]{36}"),

            # Slack Tokens
            "slack_token": re.compile(r"xox[baprs]-([0-9a-zA-Z]{10,48})"),

            # JWT Tokens
            "jwt_token": re.compile(r"eyJ[a-zA-Z0-9_-]*\.eyJ[a-zA-Z0-9_-]*\.[a-zA-Z0-9_-]*"),

            # RSA Private Keys
            "rsa_private_key": re.compile(r"-----BEGIN RSA PRIVATE KEY-----"),

            # EC Private Keys
            "ec_private_key": re.compile(r"-----BEGIN EC PRIVATE KEY-----"),

            # OpenSSH Private Keys
            "openssh_private_key": re.compile(r"-----BEGIN OPENSSH PRIVATE KEY-----"),

            # Generic Private Keys
            "generic_private_key": re.compile(r"-----BEGIN PRIVATE KEY-----"),

            # PostgreSQL Connection Strings
            "postgres_url": re.compile(
                r"postgres(?:ql)?://[a-zA-Z0-9_-]+:[a-zA-Z0-9_!@#$%^&*()+-]+@[a-zA-Z0-9.-]+:\d+/[a-zA-Z0-9_-]+"
            ),

            # MySQL Connection Strings
            "mysql_url": re.compile(
                r"mysql://[a-zA-Z0-9_-]+:[a-zA-Z0-9_!@#$%^&*()+-]+@[a-zA-Z0-9.-]+:\d+/[a-zA-Z0-9_-]+"
            ),

            # Generic API Keys
            "generic_api_key": re.compile(
                r"(?i)(api[_-]?key|apikey|secret[_-]?key)['\"]?\s*[:=]\s*['\"]([a-zA-Z0-9_\-]{32,})['\"]"
            ),
        }

    async def detect(
        self, file_path: Path, content: str, file_type: str | None = None
    ) -> List[Vulnerability]:
        """
        Detect hardcoded secrets in file content.

        Args:
            file_path: Path to the file
            content: File content
            file_type: File type (optional)

        Returns:
            List of detected vulnerabilities
        """
        vulnerabilities: List[Vulnerability] = []

        # Split content into lines for line number tracking
        lines = content.split("\n")

        for secret_type, pattern in self.patterns.items():
            for line_num, line in enumerate(lines, start=1):
                matches = pattern.finditer(line)

                for match in matches:
                    # Extract the matched secret
                    secret_value = match.group(0)

                    # Skip if it looks like a placeholder
                    if self._is_placeholder(secret_value):
                        continue

                    # Create vulnerability
                    vuln = Vulnerability(
                        type=VulnerabilityType.SECRET_EXPOSURE,
                        title=f"Hardcoded {self._format_secret_type(secret_type)}",
                        description=self._generate_description(secret_type, secret_value),
                        severity=self._determine_severity(secret_type),
                        confidence=self._determine_confidence(secret_type, secret_value),
                        file_path=str(file_path),
                        line_number=line_num,
                        code_snippet=line.strip(),
                        cwe_id="CWE-798",  # Use of Hard-coded Credentials
                        cvss_score=9.1,
                        remediation=self._generate_remediation(secret_type),
                        references=[
                            "https://cwe.mitre.org/data/definitions/798.html",
                            "https://owasp.org/Top10/A07_2021-Identification_and_Authentication_Failures/",
                        ],
                        detector=self.name,
                        engine="static",
                        mitre_attack_ids=["T1552.001"],  # Unsecured Credentials: Credentials In Files
                    )

                    vulnerabilities.append(vuln)

        return vulnerabilities

    def _is_placeholder(self, secret: str) -> bool:
        """Check if the secret is likely a placeholder."""
        placeholders = [
            "your_api_key",
            "your_secret",
            "placeholder",
            "example",
            "fake",
            "test",
            "dummy",
            "xxx",
            "yyy",
            "zzz",
            "000",
            "111",
        ]
        secret_lower = secret.lower()
        return any(placeholder in secret_lower for placeholder in placeholders)

    def _format_secret_type(self, secret_type: str) -> str:
        """Format secret type for display."""
        return secret_type.replace("_", " ").title()

    def _generate_description(self, secret_type: str, secret_value: str) -> str:
        """Generate vulnerability description."""
        # Redact most of the secret for security
        redacted = secret_value[:8] + "..." + ("*" * 8)

        descriptions = {
            "aws_access_key": f"AWS Access Key found: {redacted}. This grants programmatic access to AWS resources.",
            "aws_secret_key": f"AWS Secret Key found: {redacted}. This can be used to authenticate AWS API requests.",
            "openai_api_key": f"OpenAI API Key found: {redacted}. This provides access to OpenAI's API services.",
            "anthropic_api_key": f"Anthropic Claude API Key found: {redacted}. This grants access to Claude AI models.",
            "github_token": f"GitHub Personal Access Token found: {redacted}. This can access GitHub repositories and data.",
            "slack_token": f"Slack API Token found: {redacted}. This provides access to Slack workspace data.",
            "jwt_token": f"JWT Token found: {redacted}. This may contain sensitive authentication information.",
            "rsa_private_key": "RSA Private Key found. This can be used to decrypt data or impersonate the key owner.",
            "postgres_url": "PostgreSQL connection string with embedded credentials found.",
            "mysql_url": "MySQL connection string with embedded credentials found.",
        }

        return descriptions.get(
            secret_type,
            f"{self._format_secret_type(secret_type)} found: {redacted}. Hardcoded secrets pose a security risk.",
        )

    def _determine_severity(self, secret_type: str) -> Severity:
        """Determine severity based on secret type."""
        critical_types = [
            "aws_access_key",
            "aws_secret_key",
            "rsa_private_key",
            "ec_private_key",
            "openssh_private_key",
            "generic_private_key",
        ]

        high_types = [
            "openai_api_key",
            "anthropic_api_key",
            "github_token",
            "postgres_url",
            "mysql_url",
        ]

        if secret_type in critical_types:
            return Severity.CRITICAL
        elif secret_type in high_types:
            return Severity.HIGH
        else:
            return Severity.MEDIUM

    def _determine_confidence(self, secret_type: str, secret_value: str) -> Confidence:
        """Determine confidence level."""
        # Private keys have very specific formats - high confidence
        if "private_key" in secret_type:
            return Confidence.HIGH

        # Well-known API key formats - high confidence
        high_confidence_types = [
            "aws_access_key",
            "openai_api_key",
            "anthropic_api_key",
            "github_token",
        ]

        if secret_type in high_confidence_types:
            return Confidence.HIGH

        # Generic patterns - medium confidence
        return Confidence.MEDIUM

    def _generate_remediation(self, secret_type: str) -> str:
        """Generate remediation advice."""
        if "api_key" in secret_type or "token" in secret_type:
            return (
                "1. Immediately revoke this API key/token\n"
                "2. Remove the hardcoded credential from the source code\n"
                "3. Use environment variables or a secrets manager (e.g., HashiCorp Vault, AWS Secrets Manager)\n"
                "4. Rotate all credentials that may have been exposed\n"
                "5. Review access logs for unauthorized usage"
            )

        elif "private_key" in secret_type:
            return (
                "1. Revoke and regenerate this private key immediately\n"
                "2. Remove the private key from source control\n"
                "3. Store private keys in a secure location with restricted access\n"
                "4. Use a secrets manager or encrypted storage\n"
                "5. Review git history and remove key from all commits"
            )

        elif "url" in secret_type:
            return (
                "1. Change the database password immediately\n"
                "2. Remove the connection string from source code\n"
                "3. Use environment variables for database credentials\n"
                "4. Review database access logs for unauthorized connections\n"
                "5. Consider using IAM authentication for cloud databases"
            )

        else:
            return (
                "1. Remove the hardcoded secret from source code\n"
                "2. Use environment variables or a secrets manager\n"
                "3. Rotate the exposed credential\n"
                "4. Implement secret scanning in CI/CD pipeline"
            )

File: rust-legacy/src/detectors/secrets.rs
//! Secrets detection

use anyhow::Result;
use once_cell::sync::Lazy;
use regex::Regex;
use std::collections::HashMap;

use crate::models::vulnerability::{Location, Severity, Vulnerability, VulnerabilityType};

/// Secret pattern definition
struct SecretPattern {
    name: &'static str,
    regex: Regex,
    description: &'static str,
}

/// All secret patterns we scan for
static SECRET_PATTERNS: Lazy<Vec<SecretPattern>> = Lazy::new(|| {
    vec![
        // AWS Access Keys
        SecretPattern {
            name: "AWS Access Key ID",
            regex: Regex::new(r#"(?i)(AKIA[A-Z0-9]{16})"#).unwrap(),
            description: "AWS Access Key ID detected",
        },
        SecretPattern {
            name: "AWS Secret Access Key",
            regex: Regex::new(r#"(?i)(ASIA[A-Z0-9]{16})"#).unwrap(),
            description: "AWS Session Token detected",
        },
        // OpenAI API Keys
        SecretPattern {
            name: "OpenAI API Key",
            regex: Regex::new(r#"(sk-[a-zA-Z0-9]{20}T3BlbkFJ[a-zA-Z0-9]{20})"#).unwrap(),
            description: "OpenAI API key detected",
        },
        SecretPattern {
            name: "OpenAI API Key (Legacy)",
            regex: Regex::new(r#"(sk-[a-zA-Z0-9]{48})"#).unwrap(),
            description: "OpenAI API key (legacy format) detected",
        },
        // Anthropic API Keys
        SecretPattern {
            name: "Anthropic API Key",
            regex: Regex::new(r#"(sk-ant-[a-zA-Z0-9-]{95})"#).unwrap(),
            description: "Anthropic (Claude) API key detected",
        },
        // JWT Tokens
        SecretPattern {
            name: "JWT Token",
            regex: Regex::new(r#"(eyJ[A-Za-z0-9_-]{10,}\.eyJ[A-Za-z0-9_-]{10,}\.[A-Za-z0-9_-]{10,})"#).unwrap(),
            description: "JWT token detected",
        },
        // Private Keys
        SecretPattern {
            name: "RSA Private Key",
            regex: Regex::new(r#"-----BEGIN (RSA |EC |OPENSSH )?PRIVATE KEY-----"#).unwrap(),
            description: "Private key detected",
        },
        // Database Connection Strings
        SecretPattern {
            name: "PostgreSQL Connection String",
            regex: Regex::new(r#"postgres://[^:]+:[^@]+@[^/]+/\S+"#).unwrap(),
            description: "PostgreSQL connection string with credentials detected",
        },
        SecretPattern {
            name: "MySQL Connection String",
            regex: Regex::new(r#"mysql://[^:]+:[^@]+@[^/]+/\S+"#).unwrap(),
            description: "MySQL connection string with credentials detected",
        },
        // GitHub Tokens
        SecretPattern {
            name: "GitHub Token",
            regex: Regex::new(r#"(ghp_[a-zA-Z0-9]{36})"#).unwrap(),
            description: "GitHub Personal Access Token detected",
        },
        SecretPattern {
            name: "GitHub OAuth Token",
            regex: Regex::new(r#"(gho_[a-zA-Z0-9]{36})"#).unwrap(),
            description: "GitHub OAuth Token detected",
        },
        // Generic API Keys (high entropy strings)
        SecretPattern {
            name: "Generic API Key",
            regex: Regex::new(r#"(?i)(api[_-]?key|apikey|api[_-]?secret)['"\s]*[:=]\s*['"]?([a-zA-Z0-9_\-]{32,})['"]?"#).unwrap(),
            description: "Generic API key detected",
        },
        // Slack Tokens
        SecretPattern {
            name: "Slack Token",
            regex: Regex::new(r#"(xox[pborsa]-[0-9]{10,13}-[0-9]{10,13}-[0-9]{10,13}-[a-z0-9]{32})"#).unwrap(),
            description: "Slack token detected",
        },
        // Google API Keys
        SecretPattern {
            name: "Google API Key",
            regex: Regex::new(r#"AIza[0-9A-Za-z_-]{35}"#).unwrap(),
            description: "Google API key detected",
        },
        // Generic Passwords in Code
        SecretPattern {
            name: "Hardcoded Password",
            regex: Regex::new(r#"(?i)(password|passwd|pwd)['"\s]*[:=]\s*['"]([^'"\s]{8,})['"]"#).unwrap(),
            description: "Hardcoded password detected",
        },
    ]
});

/// Detect exposed secrets in code
pub fn detect(content: &str, file_path: &str) -> Result<Vec<Vulnerability>> {
    let mut vulnerabilities = Vec::new();
    let mut id_counter = 1;

    for (line_num, line) in content.lines().enumerate() {
        for pattern in SECRET_PATTERNS.iter() {
            if let Some(captures) = pattern.regex.captures(line) {
                // Get the matched secret (first capture group or entire match)
                let secret_text = captures
                    .get(1)
                    .or_else(|| captures.get(0))
                    .map(|m| m.as_str())
                    .unwrap_or("");

                // Redact the secret in the description
                let redacted = redact_secret(secret_text);

                let vuln = Vulnerability::new(
                    format!("SEC-{:03}", id_counter),
                    VulnerabilityType::SecretsLeakage,
                    Severity::Critical,
                    format!("{} Found", pattern.name),
                    pattern.description.to_string(),
                )
                .with_location(
                    Location::new(file_path)
                        .with_line(line_num + 1)
                        .with_column(line.find(secret_text).unwrap_or(0) + 1),
                )
                .with_impact(format!(
                    "Exposed {} can be used for unauthorized access",
                    pattern.name
                ))
                .with_remediation(format!(
                    "Remove {} from source code and use environment variables or secure secret management",
                    pattern.name
                ))
                .with_code_snippet(format!("{}\n{}", line, format!("Secret: {}", redacted)))
                .with_confidence(0.95);

                // Add evidence
                let mut evidence = HashMap::new();
                evidence.insert(
                    "secret_type".to_string(),
                    serde_json::json!(pattern.name),
                );
                evidence.insert(
                    "redacted_value".to_string(),
                    serde_json::json!(redacted),
                );
                let vuln = vuln.with_evidence(evidence);

                vulnerabilities.push(vuln);
                id_counter += 1;
            }
        }
    }

    Ok(vulnerabilities)
}

/// Redact a secret for safe display
fn redact_secret(secret: &str) -> String {
    if secret.len() <= 8 {
        "*".repeat(secret.len())
    } else {
        let prefix_len = 4.min(secret.len() / 4);
        let suffix_len = 4.min(secret.len() / 4);
        let prefix = &secret[..prefix_len];
        let suffix = &secret[secret.len() - suffix_len..];
        format!("{}...{}", prefix, suffix)
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_detect_aws_key() {
        let content = r#"
            AWS_ACCESS_KEY_ID = "AKIAIOSFODNN7EXAMPLE"
            AWS_SECRET_ACCESS_KEY = "wJalrXUtnFEMI/K7MDENG/bPxRfiCYEXAMPLEKEY"
        "#;

        let vulns = detect(content, "test.py").unwrap();
        assert!(!vulns.is_empty());
        assert!(vulns.iter().any(|v| v.title.contains("AWS")));
    }

    #[test]
    fn test_detect_openai_key() {
        let content = r#"
            OPENAI_API_KEY = "sk-1234567890abcdefghijklmnopqrstuvwxyz1234567890"
        "#;

        let vulns = detect(content, "test.py").unwrap();
        assert!(!vulns.is_empty());
        assert!(vulns.iter().any(|v| v.title.contains("OpenAI")));
    }

    #[test]
    fn test_detect_private_key() {
        let content = r#"
            -----BEGIN RSA PRIVATE KEY-----
            MIIEpAIBAAKCAQEA...
        "#;

        let vulns = detect(content, "test.py").unwrap();
        assert!(!vulns.is_empty());
        assert!(vulns.iter().any(|v| v.title.contains("Private Key")));
    }

    #[test]
    fn test_redact_secret() {
        let short = redact_secret("short");
        assert!(short.contains("*"));
        assert_eq!(redact_secret("AKIAIOSFODNN7EXAMPLE"), "AKIA...MPLE");
        assert_eq!(redact_secret("verylongsecretkey12345678"), "very...5678");
    }
}

